test.cc

```diff
@@ -275,3 +277,192 @@ TEST(Sherlock, SubscribeToStreamViaHTTP) {
+
+  KeyValueEntry() = default;
+  KeyValueEntry(const int key, const double value) : key_(key), value_(value) {}
+
+  const IntKey& key() const { return key_; }
+  void set_key(const IntKey& key) { key_ = key; }
+
+  template <typename A>
+  void serialize(A& ar) {
+    ar(CEREAL_NVP(key_), CEREAL_NVP(value_));
+  }
+};
+
+struct KeyValueSubscriptionData {
+  atomic_size_t seen_;
+  string results_;
+  KeyValueSubscriptionData() : seen_(0u) {}
+};
+
+struct KeyValueAggregateListener {
+  KeyValueSubscriptionData& data_;
+  size_t max_to_process_ = static_cast<size_t>(-1);
+
+  KeyValueAggregateListener() = delete;
+  explicit KeyValueAggregateListener(KeyValueSubscriptionData& data) : data_(data) {}
+
+  KeyValueAggregateListener& SetMax(size_t cap) {
+    max_to_process_ = cap;
+    return *this;
+  }
+
+  bool Entry(const KeyValueEntry& entry, size_t index, size_t total) {
+    static_cast<void>(index);
+    static_cast<void>(total);
+    if (data_.seen_) {
+      data_.results_ += ",";
+    }
+    data_.results_ += Printf("%d=%.2lf", entry.key(), entry.value_);
+    ++data_.seen_;
+    return data_.seen_ < max_to_process_;
+  }
+
+  void Terminate() {
+    if (data_.seen_) {
+      data_.results_ += ",";
+    }
+    data_.results_ += "DONE";
+  }
+};
+
+TEST(Sherlock, NonPolymorphicKeyValueStorage) {
+  typedef sherlock::kv_storage::API<KeyValueEntry> KVS;
+  KVS api("non_polymorphic_kv_storage");
+
+  // Add the first key-value pair.
+  // Use `UnsafeStream()`, since generally the only way to access the underlying stream is to make API calls.
+  api.UnsafeStream().Emplace(2, 0.5);
+
+  while (!api.CaughtUp()) {
+    // Spin lock, for the purposes of this test.
+    // Ensure that the data has reached the the processor that maintains the in-memory state of the API.
+  }
+
+  // Future expanded syntax.
+  std::future<KeyValueEntry> f1 = api.AsyncGet(KVS::T_KEY(2));
+  KeyValueEntry r1 = f1.get();
+  EXPECT_EQ(0.5, r1.value_);
+
+  // Future short syntax.
+  EXPECT_EQ(0.5, api.AsyncGet(KVS::T_KEY(2)).get().value_);
+
+  // Callback version.
+  struct CallBackTest {
+    explicit CallBackTest(int key, double value, bool expect_fail = false) :
+        key(key), value(value), expect_fail(expect_fail) {}
+
+    void found(const KeyValueEntry& entry) {
+      called = true;
+      EXPECT_FALSE(expect_fail);
+      EXPECT_EQ(key, entry.key()());
+      EXPECT_EQ(value, entry.value_);
+    }
+    void not_found(const IntKey& key) {
+      called = true;
+      EXPECT_TRUE(expect_fail);
+      EXPECT_EQ(this->key, key());
+    }
+    void added() {
+      called = true;
+      EXPECT_FALSE(expect_fail);
+    }
+    void already_exists() {
+      called = true;
+      EXPECT_TRUE(expect_fail);
+    }
+
+    const int key;
+    const double value;
+    const bool expect_fail;
+    bool called = false;
+  };
+
+  CallBackTest cbt1(2, 0.5);
+  auto cbf1 = [&](const KeyValueEntry& entry) { cbt1.found(entry); };
```

Comment:
`const auto` is a slightly better syntax.

In your case though, `std::bind` + `std::placeholders::_1` is probably easier to read -- and saves on the variable!